### PR TITLE
Vcf Info/Format Parsing

### PR DESCRIPTION
**Pull request by dasnellings**
This PR focuses on using the header information to parse the info and format fields in a vcf as well as functions to query the parsed fields. This is not done by default in the vcf struct as it can have a significant cost proportional to the size of the info/format fields. Instead, these fields are first initialized by calling `ParseInfo` and `ParseFormat` to initialize unexported maps in the Vcf struct. These maps can then be queried by typed query functions (e.g. `QueryInt`).

### CODE REVIEW

File: vcf/queryInfo_test.go [hunk 1]
```diff
@@ -0,0 +1,255 @@
+package vcf
+
+import (
+	"testing"
+)
+
+var expectedFormat = map[string]interface{}{
+	"FormatF": [][]int{{1}, {2}},
+	"FormatJ": [][]rune{{'F', 'E'}, {'D', 'A'}},
+	"FormatK": [][]string{{"the", "quick", "brown", "fox"}, {"jumps", "over", "the", "lazy", "dog"}},
```
Comment (craiglowe) [inline vcf/queryInfo_test.go:10]: ha ha, I remember typing this so many times in middle school

**Review by craiglowe — COMMENTED** [review]
Leaving a couple comments.  This looks like some really nice code.
Thank you for breaking this into some bitesize pull requests.
Is there a program coming where this was needed functionality?  I could imagine this being the backbone of some really nice functionality to filter vcf lines.  A program like that could also really showcase how to use the functionality.

File: vcf/queryInfo.go [hunk 2]
```diff
@@ -0,0 +1,314 @@
...
+func ParseInfo(v Vcf, header Header) Vcf {
+	v.parsedInfo = make(map[string]interface{})
+	if v.Info == "." {
+		return v
+	}
+	var fields, tagValuePair []string
+	fields = strings.Split(v.Info, ";")
+	for i := range fields {
+		tagValuePair = strings.Split(fields[i], "=")
+		if len(tagValuePair) > 2 {
```
[the remark this answers is not in the document]
Comment (dasnellings) [reply vcf/queryInfo.go:22]: Info can be either empty `"empty->;;<-empty"` (len == 0), 
a flag `"flag->;TrueFlag;<-flag"` (len == 1, a boolean presence/absence), 
or a value delimited by "=" `"value->;Val=1;<-value"` (len == 2). 

As a literal "=" is illegal in the info field, anything >2 is invalid.

File: vcf/vcf.go [hunk 3]
```diff
@@ -26,7 +39,6 @@ type GenomeSample struct {
 	FormatData []string // FormatData contains additional sample fields after the genotype, which are parsed into a slice delimited by colons. Currently contains a dummy empty string in FormatData[0] corresponding to "GT" in Format, so indices in FormatData will match the indices in Format.
 }
 
-// Header contains all of the information present in the header section of a VCf, delimited by line to form a slice of strings.
-type Header struct {
-	Text []string
+func (v Vcf) String() string {
```
[the remark this answers is not in the document]
Comment (dasnellings) [reply vcf/vcf.go:42]: Yes, done.

**Review by dasnellings — COMMENTED** [review]
I need to filter some vcfs by info and format fields so this is the infrastructure to start doing that. My next plan is to add this functionality to the vcfFilter cmd.

**Review by RavenYL — APPROVED** [review]
Excited to use this on vcfs!

**Review by craiglowe — APPROVED** [review]
Great Work!  Wonderful to see this!

File: vcf/queryInfo.go [hunk 4]
```diff
@@ -0,0 +1,314 @@
...
+		}
+		if length != num {
+			return false
+		}
+	}
+	return true
+}
+
+// QueryInt retrieves integer values stored in the Info or Format fields of a vcf record.
+// The input is a Key struct which is retrieved from the header and is keyed by Id.
```
Comment (cshoben) [inline vcf/queryInfo.go:191]: Key is used above but not defined until this line.

File: vcf/queryInfo.go [hunk 5]
```diff
@@ -0,0 +1,314 @@
...
+	interfValue := query(v, k)      // query value and store resulting interface
+	value, ok := interfValue.(bool) // assert value to the expected type
+	if !ok {                        // panic if interfValue type does not match the expected type
+		log.Panicf("value for tag '%s' in vcf at position '%d' "+
+			"is not a bool as expected by header", k.Id, v.Pos)
+	}
+	return value
+}
+
+// QueryString retrieves string values stored in the Info or Format fields of a vcf record.
```
Comment (cshoben) [inline vcf/queryInfo.go:254]: I really appreciate these descriptions!

File: vcf/queryInfo_test.go [hunk 6]
```diff
@@ -0,0 +1,255 @@
...
+	if !ok {
+		return false
+	}
+	if aBool != bBool {
+		return false
+	}
+	return true
+}
+
+func TestQueryInt(t *testing.T) {
```
Comment (cshoben) [inline vcf/queryInfo_test.go:198]: Is this meant to be QueryInfo? Or maybe QueryInt is on purpose?

**Review by cshoben — APPROVED** [review]
A few minor comments. I really appreciate your explicit descriptions before the functions.

**Pull request merged**